pulsar-common/src/main/java/org/apache/pulsar/common/util/SslContextRefresher.java

```diff
@@ -0,0 +1,146 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.security.GeneralSecurityException;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Sets;
+
+import io.netty.handler.ssl.SslContext;
+
+public class SslContextRefresher implements Runnable {
+    private final boolean tlsAllowInsecureConnection;
+    private final String tlsTrustCertsFilePath;
+    private final String tlsCertificateFilePath;
+    private final String tlsKeyFilePath;
+    private final Set<String> tlsCiphers;
+    private final Set<String> tlsProtocols;
+    private final boolean tlsRequireTrustedClientCertOnConnect;
+    private final ScheduledExecutorService eventLoopGroup;
+    private final Set<String> files;
+    private final Set<WatchKey> watchKeys;
+    private final long delay;
+    private final TimeUnit timeunit;
+    private volatile SslContext sslContext;
+
+    public SslContextRefresher(boolean allowInsecure, String trustCertsFilePath, String certificateFilePath,
+            String keyFilePath, Set<String> ciphers, Set<String> protocols, boolean requireTrustedClientCertOnConnect,
+            ScheduledExecutorService eventLoopGroup, long delay, TimeUnit timeunit)
+            throws SSLException, FileNotFoundException, GeneralSecurityException, IOException {
+        this.tlsAllowInsecureConnection = allowInsecure;
+        this.tlsTrustCertsFilePath = trustCertsFilePath;
+        this.tlsCertificateFilePath = certificateFilePath;
+        this.tlsKeyFilePath = keyFilePath;
+        this.tlsCiphers = ciphers;
+        this.tlsProtocols = protocols;
+        this.tlsRequireTrustedClientCertOnConnect = requireTrustedClientCertOnConnect;
+        this.watchKeys = Sets.newHashSet();
+        this.files = Sets.newHashSet();
+        this.delay = delay;
+        this.timeunit = timeunit;
+        this.eventLoopGroup = eventLoopGroup;
+
+        files.add(trustCertsFilePath);
+        files.add(certificateFilePath);
+        files.add(keyFilePath);
+
+        // Remove all Null Or Empty Files
+        files.removeIf(file -> Strings.isNullOrEmpty(file));
+
+        this.sslContext = SecurityUtility.createNettySslContextForServer(tlsAllowInsecureConnection,
+                tlsTrustCertsFilePath, tlsCertificateFilePath, tlsKeyFilePath, tlsCiphers, tlsProtocols,
+                tlsRequireTrustedClientCertOnConnect);
+
+        for (String file : files) {
+            LOG.info("File {}", file);
+            Path p = Paths.get(file).getParent();
+            WatchService watchService = p.getFileSystem().newWatchService();
```

Comment:
Why to do we need to poll if we're using a "watch-service"? Isn't this supposed to notify when the files are changed?

Reply:
I believe Watch service doesn't have a callback, it just has a polling mechanism.
https://docs.oracle.com/javase/8/docs/api/java/nio/file/WatchService.html